fili-core/src/main/java/com/yahoo/bard/webservice/web/filterparser/ApiFiltersListListener.java

```diff
@@ -0,0 +1,219 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import com.yahoo.bard.webservice.config.BardFeatureFlag;
+import com.yahoo.bard.webservice.data.dimension.Dimension;
+import com.yahoo.bard.webservice.data.dimension.DimensionDictionary;
+import com.yahoo.bard.webservice.data.dimension.DimensionField;
+import com.yahoo.bard.webservice.table.LogicalTable;
+import com.yahoo.bard.webservice.web.ApiFilter;
+import com.yahoo.bard.webservice.web.BadApiRequestException;
+import com.yahoo.bard.webservice.web.BadFilterException;
+import com.yahoo.bard.webservice.web.ErrorMessageFormat;
+import com.yahoo.bard.webservice.web.FilterOperation;
+
+import static com.yahoo.bard.webservice.web.ErrorMessageFormat.FILTER_DIMENSION_NOT_IN_TABLE;
+import static com.yahoo.bard.webservice.web.ErrorMessageFormat.FILTER_DIMENSION_UNDEFINED;
+import static com.yahoo.bard.webservice.web.ErrorMessageFormat.FILTER_FIELD_NOT_IN_DIMENSIONS;
+import static com.yahoo.bard.webservice.web.ErrorMessageFormat.FILTER_OPERATOR_INVALID;
+
+/**
+ * Listener used to walk the Filters parse tree and build ApiFilter objects.
+ */
+public class ApiFiltersListListener extends FiltersBaseListener {
+    private static final Logger LOG = LoggerFactory.getLogger(ApiFiltersListListener.class);
+
+    private final Map<FiltersParser.FilterComponentContext, ApiFilter> filters = new HashMap<> ();
+    private final Map<Dimension, Set<ApiFilter>> dimensionFiltersMap = new LinkedHashMap<>();
+    private final List<Exception> errors = new LinkedList<>();
+    private final LogicalTable table;
+    private final DimensionDictionary dimensionDictionary;
+
+    /**
+     * Constructor.
+     *
+     * @param table logical table that the filter applies to
+     * @param dimensionDictionary dictionary of all the valid dimensions
+     */
+    public ApiFiltersListListener(LogicalTable table, DimensionDictionary dimensionDictionary) {
+        this.table = table;
+        this.dimensionDictionary = dimensionDictionary;
+    }
+
+    @Override
+    public void exitFilters(@NotNull FiltersParser.FiltersContext ctx) {
+        for (FiltersParser.FilterComponentContext f : ctx.filterComponent()) {
+            ApiFilter filter = filters.get(f);
+            if (filter == null) {
+                // there was some error and it has been saved in the errors list
+                break;
+            }
+            Dimension dimension = filter.getDimension();
+            if (!dimensionFiltersMap.containsKey(dimension)) {
+                dimensionFiltersMap.put(dimension, new LinkedHashSet<>());
+            }
+            Set<ApiFilter> filterSet = dimensionFiltersMap.get(dimension);
+            filterSet.add(filter);
+            dimensionFiltersMap.put(dimension, filterSet);
```

Comment:
Everything from the if statement on can be consolidated into a single statement with `computeIfAbsent`:

```java
dimensionFiltersMap.computeIfAbsent(filter.getDimension(), ignored -> new LinkedHashSet<>()).add(filter);
```